File: tests/fixtures/lang_smoke/go/shapes/shapes.go
// Package shapes: interface + implementor (IMPLEMENTS), struct + method (CONTAINS),
// method call (CALLS), and a parameter used downstream (data_flow).
package shapes

// Shape is implemented by Rectangle -> IMPLEMENTS edge.
type Shape interface {
	Area() float64
	ScaledArea(factor float64) float64
}

// Rectangle is a struct with methods -> CONTAINS edges.
type Rectangle struct {
	Width  float64
	Height float64
}

// Area returns the rectangle's area.
func (r Rectangle) Area() float64 {
	return r.Width * r.Height
}

// ScaledArea multiplies the area by factor.
func (r Rectangle) ScaledArea(factor float64) float64 {
	// `factor` flows downstream (data_flow); r.Area() is a CALLS edge.
	base := r.Area()
	return base * factor
}

File: tests/fixtures/lang_smoke/rust/shapes.rs
// Trait + implementor (IMPLEMENTS/EXTENDS), struct + method (CONTAINS),
// method call (CALLS), and a parameter used downstream (data_flow).

/// Shape is implemented by Rectangle -> IMPLEMENTS edge.
pub trait Shape {
    fn area(&self) -> f64;
    fn scaled_area(&self, factor: f64) -> f64;
}

/// Rectangle is a struct with methods -> CONTAINS edges.
pub struct Rectangle {
    pub width: f64,
    pub height: f64,
}

impl Shape for Rectangle {
    fn area(&self) -> f64 {
        self.width * self.height
    }

    fn scaled_area(&self, factor: f64) -> f64 {
        // `factor` flows downstream (data_flow); self.area() is a CALLS edge.
        let base = self.area();
        base * factor
    }
}
